core/services/blockhashstore/feeder.go

```diff
@@ -0,0 +1,164 @@
+package blockhashstore
+
+import (
+	"context"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+)
+
+// Event contains metadata about a VRF randomness request or fulfillment.
+type Event struct {
+	// ID of the relevant VRF request. For a VRF V1 request, this will an encoded 32 byte array.
+	// For VRF V2, it will be an integer in string form.
+	ID string
+
+	// Block that the request or fulfillment was included in.
+	Block uint64
+}
+
+// Coordinator defines an interface for fetching request and fulfillment metadata from a VRF
+// coordinator.
+type Coordinator interface {
+	// Requests fetches VRF requests that occurred within the specified blocks.
+	Requests(ctx context.Context, fromBlock uint64, toBlock uint64) ([]Event, error)
+
+	// Fulfillments fetches VRF fulfillments that occured since the specified block.
+	Fulfillments(ctx context.Context, fromBlock uint64) ([]Event, error)
+}
+
+// BHS defines an interface for interacting with a BlockhashStore contract.
+type BHS interface {
+	// Store the hash associated with blockNum.
+	Store(ctx context.Context, blockNum uint64) error
+
+	// IsStored checks whether the hash associated with blockNum is already stored.
+	IsStored(ctx context.Context, blockNum uint64) (bool, error)
+}
+
+// NewFeeder creates a new Feeder instance.
+func NewFeeder(
+	logger logger.Logger,
+	coordinator Coordinator,
+	bhs BHS,
+	waitBlocks int,
+	lookbackBlocks int,
+	latestBlock func(ctx context.Context) (uint64, error),
+) *Feeder {
+	return &Feeder{
+		logger:         logger,
+		coordinator:    coordinator,
+		bhs:            bhs,
+		waitBlocks:     waitBlocks,
+		lookbackBlocks: lookbackBlocks,
+		latestBlock:    latestBlock,
+		stored:         make(map[uint64]struct{}),
+		lastRunBlock:   0,
+	}
+}
+
+// Feeder checks recent VRF coordinator events and stores any blockhashes for blocks within
+// waitBlocks and lookbackBlocks that have unfulfilled requests.
+type Feeder struct {
+	logger         logger.Logger
+	coordinator    Coordinator
+	bhs            BHS
+	waitBlocks     int
+	lookbackBlocks int
+	latestBlock    func(ctx context.Context) (uint64, error)
+
+	stored       map[uint64]struct{}
+	lastRunBlock uint64
+}
+
+// Run the feeder.
+func (f *Feeder) Run(ctx context.Context) {
```

Comment:
I think it might be better for `Run` to return `error` so that the caller can know if an error was encountered or not. Right now in the caller we log "BHS feeder run completed" which kind of sounds like success, even if it potentially failed for some reason (e.g, timeout on a get logs call)

Reply:
As a plus, it'll help testing these error cases by doing `require.Error(...)` in unit tests.

Reply:
Good call, done